File: PC_L4Server/Request.cpp
#include "Request.h"

Request Request::readRequest(std::vector<char> message)
{
    uint32_t command;
    bitsToInt(command, (unsigned char*) message.data() + MESSAGE_MAP::command, false);
    uint32_t rows, columns;
    bitsToInt(rows, (unsigned char*)message.data() + MESSAGE_MAP::rows, false);
    bitsToInt(columns, (unsigned char*)message.data() + MESSAGE_MAP::columns, false);
    int32_t min, max;
    bitsToInt(min, (unsigned char*)message.data() + MESSAGE_MAP::min, false);
    bitsToInt(max, (unsigned char*)message.data() + MESSAGE_MAP::max, false);

    Request req;
    req._command = static_cast<COMMAND_TYPES>(command);
    req._max = max;
    req._min = min;
    req._data = std::vector<std::vector<int32_t>>(rows, std::vector<int32_t>(columns, 0));

    int itterator = 0;
    for (uint32_t i = 0; i < rows; i++)
    {
        for (uint32_t j = 0; j < columns; j++)
        {
            bitsToInt(req._data[i][j], (unsigned char*)message.data() + MESSAGE_MAP::data + itterator, false);
            itterator += 4;
        }
    }

    return req;
}

template <typename T>
void _insertBytesToMessage(std::vector<char>& message, const T& varToInsert, const unsigned int sizeOfVar)
{
    std::vector<char> varBytes(sizeOfVar);
    intToBits<T>(varToInsert, (unsigned char*)varBytes.data(), false, sizeOfVar);
    message.insert(std::end(message), std::begin(varBytes), std::end(varBytes));
}

std::vector<char> Request::to_bytes()
{
    uint32_t rows = 0, columns = 0;

    if (_data.size() != 0)
    {
        rows = _data.size();
        columns = _data[0].size();
    }

    std::vector<char> dataBytes;

    unsigned int size = sizeof(_data[0][0]);
    for (auto row : _data)
    {
        for (auto cell : row)
        {
            std::vector<char> cellBytes(size, 0);
            intToBits(cell, (unsigned char*)cellBytes.data(), false);
            dataBytes.insert(std::end(dataBytes), std::begin(cellBytes), std::end(cellBytes));
        }
    }

    std::vector<char> message(sizeof(uint32_t), 0);

    _insertBytesToMessage(message, (uint32_t)_command, 4);
    _insertBytesToMessage(message, rows, 4);
    _insertBytesToMessage(message, columns, 4);
    _insertBytesToMessage(message, _min, 4);
    _insertBytesToMessage(message, _max, 4);

    message.insert(std::end(message), std::begin(dataBytes), std::end(dataBytes));

    std::vector<char> varBytes(4);
    intToBits(message.size(), (unsigned char*)varBytes.data(), false, size = 4);

    std::copy(std::begin(varBytes), std::end(varBytes), std::begin(message));

    return message;
}

std::ostream& operator<<(std::ostream& os, Request const& m)
{

    std::string data_to_s;

    for (auto row : m.data())
    {
        for (auto cell : row)
        {
            auto string = std::to_string(cell);
            string.resize(10,' ');
            data_to_s += string;
        }
        data_to_s += "\n";
    }


    return os << "Command: " << m.command()
        << "\nMin: " << m.getMin()
        << "\nMax: " << m.getMax()
        << "\nData:\n"
        << data_to_s;
}

File: PC_L4Client/request.py
from enum import Enum
import math
import sys

class COMMAND_TYPES(Enum):
    unknown = 0
    upload = 1
    start = 3
    status = 4

    r_idle = 100
    r_data_accepted = 101
    r_processing = 102
    r_success = 103
    r_error = 104

class MESSAGE_MAP(Enum):
    messageLength = 0
    command = 4
    rows = 8
    columns = 12
    min = 16
    max = 20
    data = 24

class request:
    command = 0
    min = 0
    max = 0
    data = 0

    def __init__(self):
        self.command = 0
        self.min = 0
        self.max = 0
        self.data = []

    def __str__(self):
        data_str = ''
        for row in self.data:
            for cell in row:
                data_str += str(cell).ljust(10)[:10]
            data_str +="\n"
        return "Command: " + str(self.command.name) + "\nMin: " + str(self.min) + "\nMax: " + str(self.max) + "\nData:\n"+ data_str

    def __bytes__(self):
        rows = 0
        columns = 0
        if len(self.data) != 0:
            rows = len(self.data)
            columns = len(self.data[0])
        
        data_bytes = bytearray(b'')
        for row in self.data:
            for cell in row:
                data_bytes.extend(cell.to_bytes(4,'big')) # here number depends on matrix dtype
        
        message = bytearray(int(0).to_bytes(4,'big') + self.command.value.to_bytes(4,'big') + rows.to_bytes(4,'big') + columns.to_bytes(4,'big') + self.min.to_bytes(4,'big') + self.max.to_bytes(4,'big'))
        message.extend(data_bytes)
        
        message[0:4] = bytearray(len(message).to_bytes(4,'big'))

        return bytes(message)

def read_request(message: bytes) -> request:
    command = int.from_bytes(message[MESSAGE_MAP.command.value:MESSAGE_MAP.command.value+4],'big')
    rows = int.from_bytes(message[MESSAGE_MAP.rows.value:MESSAGE_MAP.rows.value+4],'big')
    columns = int.from_bytes(message[MESSAGE_MAP.columns.value:MESSAGE_MAP.columns.value+4],'big')
    min_val = int.from_bytes(message[MESSAGE_MAP.min.value:MESSAGE_MAP.min.value+4],'big')
    max_val = int.from_bytes(message[MESSAGE_MAP.max.value:MESSAGE_MAP.max.value+4],'big')

    req = request()
    req.command = COMMAND_TYPES(command)
    req.max = max_val
    req.min = min_val
    req.data = [[0] * columns for _ in range(rows)]

    iterator = 0
    for i in range(rows):
        for j in range(columns):
            req.data[i][j] = int.from_bytes(message[MESSAGE_MAP.data.value + iterator:MESSAGE_MAP.data.value + iterator+4],'big')
            iterator += 4

    return req

def upload_request(matrix, min, max):
    req = request();
    req.command = COMMAND_TYPES.upload
    req.data = matrix
    req.min = min
    req.max = max
    return req

def start_request():
    req = request();
    req.command = COMMAND_TYPES.start
    return req

def status_request():
    req = request();
    req.command = COMMAND_TYPES.status
    return req


